File: pizza_bot.jl
using Printf
using HTTP
using JSON

#include("white_backend.jl")
include("psql_backend.jl")
using Main.Backend

url = "https://api.telegram.org/"
key=ENV["TELEGRAM_KEY"]

function send_message(params)
    req = HTTP.request("POST",string(url,key,"/sendMessage"),["Content-Type" => "application/json"],JSON.json(params))
end

function send_welcome(contact)
    id = contact["id"]
    name = contact["first_name"]
    if "language_code" in keys(contact)
        lc = contact["language_code"]
    else
        lc = "it"
    end
    if lc=="it"
        params = Dict("chat_id"=>id, "text"=>string("Ciao ", name, "! Sono Pizzabot\n
        Ti aiuterò a preparare pizze e focacce favolose!
        Digita /cuciniamo per iniziare e poi segui le istruzioni.

        Buon lavoro e buon appetito!!"))
    else
        params = Dict("chat_id"=>id, "text"=>string("Welcome ", name, " \nClick /cuciniamo to begin.."))
    end
    send_message(params)
end

function ask_program(contact)
    id = contact["id"]
    buttons = [Dict("text"=>"pizza"), Dict("text"=>"focaccia")]
    reply_markup = Dict("keyboard"=>[buttons], "one_time_keyboard"=>true)
    params = Dict("chat_id"=>id, "text"=>"Cosa vuoi cucinare? (pizza o focaccia)", "reply_markup"=>reply_markup)
    send_message(params)
end

function ask_people(message)
    id = message["from"]["id"]
    program = message["text"]

    if !(program  in ["pizza" "focaccia"])
        program = "pizza"
    end

    params = Dict("chat_id"=>id, "text"=>string("Bene allora ti darò istruzioni per fare la ", program))
    send_message(params)
    params = Dict("chat_id"=>id, "text"=>string("per quante persone vuoi fare la ", program, "?"))
    send_message(params)
    return program
end

function read_people(message, state)
    id = message["from"]["id"]
    people = message["text"]
    program = get_key(id, "program")
    try
        people = parse(Int, people)
        if people <= 0
            throw("too few people")
        end
        send_recipe(message, people)
        return state + 1
    catch
        params = Dict("chat_id"=>id, "text"=>string("Non ho capito. Per quante persone?"))
        send_message(params)
        return state
    end
end

function send_recipe(message, people)
    id = message["from"]["id"]
    program = get_key(id, "program")
    if program == "pizza"
        send_pizza(id, people)
    elseif program == "focaccia"
        send_focaccia(id, people)
    end
end

function send_pizza(id, people)
    pizza_recipe = [1000 100 500]
    recipe = pizza_recipe / 4. * people
    params = Dict("chat_id"=>id, "text"=>string("Ecco le dosi per $(people) persone:\n
    $(@sprintf("%.1f", recipe[1]/1000)) Kg di farina\n
    $(@sprintf("%.1f", recipe[2])) ml di olio\n
    $(@sprintf("%.1f", recipe[3])) ml di acqua\n
    lievito q.b.\n
    sale q.b.\n"))
    send_message(params)
end

function send_focaccia(id, people)
    focaccia_recipe = [660 50 440]
    recipe = focaccia_recipe / 4. * people

    number, rectangular_str, circular_str, rectangular, circular = get_teglia(sum(recipe))


    params = Dict("chat_id"=>id, "text"=>string("Ecco le dosi per $(people) persone:\n
    $(@sprintf("%.3f", recipe[1]/1000)) Kg di farina di Manitoba (W=280)\n
    $(@sprintf("%.1f", recipe[2])) g di olio\n
    $(@sprintf("%.1f", recipe[3])) ml di acqua\n
    lievito q.b.\n
    sale q.b.\n\n
    Puoi usare $(circular_str) di diametro: $(@sprintf("%.1f", circular))cm\n
    oppure $(rectangular_str) $(@sprintf("%.1f", rectangular[1])) cm X $(@sprintf("%.1f", rectangular[2])) cm"))
    send_message(params)
end

function get_teglia(impasto)
    area = impasto / 0.60
    rapp = 42 / 35
    area_max = 42 * 35
    number = Int(round(floor(area/area_max) + 1))

    area = area / number # area singola teglia

    h = sqrt(area / rapp)
    b = area / h
    rectangular = [b h]
    circular = sqrt(area/pi)*2

    if number > 1
        rectangular_str = "$(@sprintf("%d", number)) teglie rettangolari"
        circular_str = "$(@sprintf("%d", number)) teglie circolari"
    else
        rectangular_str = "una teglia rettangolare"
        circular_str = "una teglia circolare"
    end

    return (number, rectangular_str, circular_str, rectangular, circular)
end

function send_bye(contact)
    id = contact["id"]
    name = contact["first_name"]
    if "language_code" in keys(contact)
        lc = contact["language_code"]
    else
        lc = "it"
    end
    if lc=="it"
        params = Dict("chat_id"=>id, "text"=>string("Buon appetito ", name, "!"))
    else
        params = Dict("chat_id"=>id, "text"=>string("Enjoy your meal ", name, "!"))
    end
    send_message(params)
end

function parse_message(message)
    id = message["message"]["from"]["id"]
    update_contact_list(message)
    state = get_state(id)

    if (state == 0) | (message["message"]["text"]=="/start")
        send_welcome(message["message"]["from"])
        set_state(id, 1)
    elseif (state == 1) | (message["message"]["text"]=="/cuciniamo")
        ask_program(message["message"]["from"])
        set_state(id, 2)
    elseif state == 2
        program = ask_people(message["message"])
        set_state(id, 3)
        set_key(id,"program",program)
    elseif state == 3
        new_state = read_people(message["message"], 3)
        state = new_state
        set_state(id, state)
    end

    if state == 4
        send_bye(message["message"]["from"])
        set_state(id, 1)
    end
end

function run()
    offset = -1

    println("ready!")
    while true
    #for i in 1:100
        if offset >= 0
            params = Dict("offset"=>offset)
        else
            params = Dict()
        end
        req = HTTP.request("GET", string(url,key,"/getUpdates"), ["Content-Type" => "application/json"],JSON.json(params))
        body = String(req.body)
        result = JSON.Parser.parse(body)

        if length(result["result"]) < 1
            continue
        end
        offset = result["result"][end]["update_id"] + 1
        for message in result["result"]
           parse_message(message)
        end
        sleep(0.1)
    end
end


init_db()
conn = get_conn()

run()

@sync close(conn)

File: serverlessApp/packages/pizzabot/getUpdate/pizza_bot.py
import json
import requests
import os, sys
import math
import time
import logging
log = logging.getLogger("pizza_bot")
hdl = logging.StreamHandler()
log.addHandler(hdl)
log.setLevel(logging.DEBUG)

from nimbella_backend import *

url = "https://api.telegram.org/"
key = os.environ["TELEGRAM_KEY"]

def send_message(params):
    req = requests.post(url+key+"/sendMessage", data = params)
    log.debug(req.json())
    log.debug(params)

def send_welcome(contact):
    id = contact["id"]
    name = contact["first_name"]
    if "language_code" in contact.keys():
        lc = contact["language_code"]
    else:
        lc = "it"

    if lc=="it":
        params = {"chat_id":id, "text":"Ciao "+name+"""! Sono Pizzabot\n
        Ti aiuterò a preparare pizze e focacce favolose!
        Digita /cuciniamo per iniziare e poi segui le istruzioni.

        Buon lavoro e buon appetito!!"""}
    else:
        params = {"chat_id":id, "text":"Welcome "+name+" \nClick /cuciniamo to begin.."}
    send_message(params)

def ask_program(contact):
    id = contact["id"]
    buttons = [[{"text":"pizza"}, {"text":"focaccia"}]]
    reply_markup = {"keyboard":buttons, "one_time_keyboard":True}
    params = {"chat_id":id, "text":"Cosa vuoi cucinare? (pizza o focaccia)", "reply_markup":json.dumps(reply_markup)}
    send_message(params)

def ask_people(message):
    id = message["from"]["id"]
    program = message["text"]

    if program not in ["pizza", "focaccia"]:
        program = "pizza"

    params = {"chat_id":id, "text":"Bene allora ti darò istruzioni per fare la "+program}
    send_message(params)
    params = {"chat_id":id, "text":"per quante persone vuoi fare la "+program+"?"}
    send_message(params)
    return program


def read_people(message, state):
    id = message["from"]["id"]
    people = message["text"]
    program = get_key(id, "program")
    try:
        people = int(people)
        if people <= 0:
            raise ValueError("too few people")
        send_recipe(message, people)
        return state + 1
    except:
        log.debug(*sys.exc_info())
        params = {"chat_id":id, "text":"Non ho capito. Per quante persone?"}
        send_message(params)
        return state
   
def send_recipe(message, people):
    id = message["from"]["id"]
    program = get_key(id, "program")
    log.debug("\n\n\n\nPROGRAM"+program)
    if "pizza" in program:
        send_pizza(id, people)
    elif "focaccia" in program:
        send_focaccia(id, people)
    
def send_pizza(id, people):
    pizza_recipe = [1000, 100, 500]
    recipe = [el / 4. * people for el in pizza_recipe]
    params = {"chat_id":id, "text":f"""Ecco le dosi per {people} persone:\n
    {recipe[0]/1000} Kg di farina\n
    {recipe[1]} ml di olio\n
    {recipe[2]} ml di acqua\n
    lievito q.b.\n
    sale q.b.\n"""}
    send_message(params)

def send_focaccia(id, people):
    focaccia_recipe = [660, 50, 440]
    recipe = [el / 4. * people for el in focaccia_recipe]
    number, rectangular_str, circular_str, rectangular, circular = get_teglia(sum(recipe))

    params = {"chat_id":id, "text":f"""Ecco le dosi per {people} persone:\n
    {recipe[0]/1000} Kg di farina di Manitoba (W=280)\n
    {recipe[1]} g di olio\n
    {recipe[2]} ml di acqua\n
    lievito q.b.\n
    sale q.b.\n\n
    Puoi usare {circular_str} di diametro: {circular}cm\n
    oppure {rectangular_str} {rectangular[0]} cm X {rectangular[1]} cm"""}
    send_message(params)

def get_teglia(impasto):
    area = impasto / 0.60
    rapp = 42 / 35
    area_max = 42 * 35
    number = int(round(math.floor(area/area_max) + 1))

    area = area / number # area singola teglia

    h = round(math.sqrt(area / rapp))
    b = round(area / h)
    rectangular = [b, h]
    circular = round(math.sqrt(area/math.pi)*2,1)

    if number > 1:
        rectangular_str = f"{number} teglie rettangolari"
        circular_str = f"{number} teglie circolari"
    else:
        rectangular_str = "una teglia rettangolare"
        circular_str = "una teglia circolare"

    return (number, rectangular_str, circular_str, rectangular, circular)

def send_bye(contact):
    id = contact["id"]
    name = contact["first_name"]
    if "language_code" in contact.keys():
        lc = contact["language_code"]
    else:
        lc = "it"
    
    if lc=="it":
        params = {"chat_id":id, "text":"Buon appetito "+name+"!"}
    else:
        params = {"chat_id":id, "text":"Enjoy your meal "+name+"!"}
    
    send_message(params)


def parse_message(message):
    id = message["message"]["from"]["id"]
    update_contact_list(message)
    state = get_state(id)

    if (state == 0) | (message["message"]["text"]=="/start"):
        send_welcome(message["message"]["from"])
        set_state(id, 1)
    elif (state == 1) | (message["message"]["text"]=="/cuciniamo"):
        ask_program(message["message"]["from"])
        set_state(id, 2)
    elif state == 2:
        program = ask_people(message["message"])
        set_state(id, 3)
        set_key(id,"program",program)
    elif state == 3:
        new_state = read_people(message["message"], 3)
        state = new_state
        set_state(id, state)

    if state == 4:
        send_bye(message["message"]["from"])
        set_state(id, 1)
   
def run():
    offset = -1

    log.info("ready!")
    while True:
    #for i in 1:100
        if offset >= 0:
            params = {"offset":offset}
        else:
            params = {}
        
        log.debug(params)

        req = requests.get(url+key+"/getUpdates", params = params)
        result = req.json()
        log.debug(result)
        if len(result["result"]) < 1:
            continue

        offset = int(result["result"][len(result["result"])-1]["update_id"]) + 1
        log.debug(f"offset {offset}")
        log.debug(result["result"])
        log.debug(len(result["result"]))
        for message in result["result"]:
           parse_message(message)
        
        time.sleep(.5)


if __name__ == "__main__":
	run()
